txnsync/txngroups_test.go

```diff
@@ -47,23 +224,287 @@ func TestTxnGroupEncoding(t *testing.T) {
+			count[txn.Txn.Type]++
+			if !txn.Sig.MsgIsZero() {
+				sigs++
+			}
+			if !txn.Msig.MsgIsZero() {
+				msigs++
+			}
+			if !txn.Lsig.MsgIsZero() {
+				lsigs++
+			}
+		}
+	}
+	fmt.Println(count)
+	fmt.Println(sigs, msigs, lsigs)
+}
+
+func BenchmarkTxnGroupEncoding(b *testing.B) {
+	txnGroups, _, _, err := txnGroupsData()
+	require.NoError(b, err)
+	var encodedGroupsBytes []byte
+
+	b.ReportAllocs()
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		encodedGroupsBytes, err := encodeTransactionGroups(txnGroups)
+		require.NoError(b, err)
+		releaseEncodedTransactionGroups(encodedGroupsBytes)
+	}
+
+	fmt.Println("new data: ", len(encodedGroupsBytes))
+}
+
+func BenchmarkTxnGroupDecoding(b *testing.B) {
+	txnGroups, genesisID, genesisHash, err := txnGroupsData()
+	require.NoError(b, err)
+
+	encodedGroupsBytes, err := encodeTransactionGroups(txnGroups)
+	require.NoError(b, err)
+
+	b.ReportAllocs()
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		_, err = decodeTransactionGroups(encodedGroupsBytes, genesisID, genesisHash)
+		require.NoError(b, err)
+	}
+}
+
+func BenchmarkTxnGroupEncodingOld(b *testing.B) {
+	txnGroups, _, _, err := txnGroupsData()
+	require.NoError(b, err)
+	var encodedGroupsBytes []byte
+
+	b.ReportAllocs()
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		encodedGroupsBytes = encodeTransactionGroupsOld(txnGroups)
+		releaseEncodedTransactionGroups(encodedGroupsBytes)
+	}
+
+	fmt.Println("old data: ", len(encodedGroupsBytes))
+}
+
+func BenchmarkTxnGroupDecodingOld(b *testing.B) {
+	txnGroups, _, _, err := txnGroupsData()
+	require.NoError(b, err)
+
+	encodedGroupsBytes := encodeTransactionGroupsOld(txnGroups)
+
+	b.ReportAllocs()
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		_, err = decodeTransactionGroupsOld(encodedGroupsBytes)
+		require.NoError(b, err)
+	}
+}
+
+// TestTxnGroupEncodingReflection generates random
+// txns of each type using reflection
+func TestTxnGroupEncodingReflection(t *testing.T) {
+	for i := 0; i < 10; i++ {
+		v0, err := protocol.RandomizeObject(&transactions.SignedTxn{})
+		require.NoError(t, err)
+		stx, ok := v0.(*transactions.SignedTxn)
+		require.True(t, ok)
+
+		var txns []transactions.SignedTxn
+		for _, txType := range protocol.TxnTypes {
+			txn := *stx
+			txn.Txn.PaymentTxnFields = transactions.PaymentTxnFields{}
+			txn.Txn.KeyregTxnFields = transactions.KeyregTxnFields{}
+			txn.Txn.AssetConfigTxnFields = transactions.AssetConfigTxnFields{}
+			txn.Txn.AssetTransferTxnFields = transactions.AssetTransferTxnFields{}
+			txn.Txn.AssetFreezeTxnFields = transactions.AssetFreezeTxnFields{}
+			txn.Txn.ApplicationCallTxnFields = transactions.ApplicationCallTxnFields{}
+			txn.Txn.CompactCertTxnFields = transactions.CompactCertTxnFields{}
+			txn.Txn.Type = txType
+			txn.Lsig.Logic = []byte("logic")
+			if i%3 != 0 {
+				txn.Sig = crypto.Signature{}
+			}
+			if i%3 != 1 {
+				txn.Msig = crypto.MultisigSig{}
+			}
+			if i%3 != 2 {
+				txn.Lsig = transactions.LogicSig{}
+			}
```

Comment:
coding style - use a switch statement, and initialize these to have some meaningful content.

Reply:
the point of this was such that if i%3 = 0, it would have a normal signature, 1 for multisig, and 2 for logicsig. I will refactor this to make it clearer.